juju/juju
# Issue error for setcharm

## Description of change

The following will attempt to throw an error if your controller is
greater than 2.5.2, yet your agents are less than and equal too
2.5.2. This prevents a potential version miss-match of LXDProfile
code attempting to use old code vs the new code. From the very
outset we tell the operator that they should upgrade.

## QA steps

Check out `2.5`:

```
export JUJU_DEV_FEATURE_FLAGS=instance-mutater
juju bootstrap lxd test
juju deploy ./testcharms/charm-repo/quantal/lxd-profile
```

Check out to this branch:

```
juju upgrade-model -m controller --build-agent=true
juju upgrade-charm lxd-profile --path ./testcharms/charm-repo/quantal/lxd-profile
```

This should return the following error:

```
Added charm "local:bionic/lxd-profile-1" to the model.
ERROR Unable to upgrade LXDProfile charms with the current model version. Please run juju upgrade-juju to upgrade the current model to match your controller.
```

To fix this, run the following:

```
juju upgrade-model
juju upgrade-charm lxd-profile --path ./testcharms/charm-repo/quantal/lxd-profile
```

This should succeed!

## Documentation changes

This will prevent upgrading of greater than 2.5.2 from units on 2.5.2
or less, to prevent conflicting code running.
